crates/services/txpool_v2/src/storage/graph.rs

```diff
@@ -0,0 +1,476 @@
+                Output::Variable { .. } => {
+                    return Err(Error::NotInsertedInputDependentOnChangeOrVariable)
+                }
+                Output::ContractCreated { .. } => {
+                    return Err(Error::NotInsertedIoContractOutput)
+                }
+            };
+        }
+        Ok(())
+    }
+
+    /// Cache the transaction information in the storage caches.
+    /// This is used to speed up the verification/dependencies searches of the transactions.
+    fn cache_tx_infos(
+        &mut self,
+        outputs: &[Output],
+        tx_id: &TxId,
+        node_id: NodeIndex,
+    ) -> Result<(), Error> {
+        for (index, output) in outputs.iter().enumerate() {
+            let index = u16::try_from(index).map_err(|_| {
+                Error::WrongOutputNumber(format!(
+                    "The number of outputs in `{}` is more than `u8::max`",
+                    tx_id
+                ))
+            })?;
+            let utxo_id = UtxoId::new(*tx_id, index);
+            match output {
+                Output::Coin { .. } | Output::Change { .. } | Output::Variable { .. } => {
+                    self.coins_creators.insert(utxo_id, node_id);
+                }
+                Output::ContractCreated { contract_id, .. } => {
+                    self.contracts_creators.insert(*contract_id, node_id);
+                }
+                _ => {}
+            }
+        }
+        Ok(())
+    }
+
+    /// Clear the caches of the storage when a transaction is removed.
+    fn clear_cache(&mut self, outputs: &[Output], tx_id: &TxId) -> Result<(), Error> {
+        for (index, output) in outputs.iter().enumerate() {
+            let index = u16::try_from(index).map_err(|_| {
+                Error::WrongOutputNumber(format!(
+                    "The number of outputs in `{}` is more than `u16::max`",
+                    tx_id
+                ))
+            })?;
+            let utxo_id = UtxoId::new(*tx_id, index);
+            match output {
+                Output::Coin { .. } | Output::Change { .. } | Output::Variable { .. } => {
+                    self.coins_creators.remove(&utxo_id);
+                }
+                Output::ContractCreated { contract_id, .. } => {
+                    self.contracts_creators.remove(contract_id);
+                }
+                _ => {}
+            }
+        }
+        Ok(())
+    }
+
+    fn get_inner(&self, index: &NodeIndex) -> Result<&StorageData, Error> {
+        self.graph
+            .node_weight(*index)
+            .ok_or(Error::TransactionNotFound(format!(
+                "Transaction with index {:?} not found",
+                index
+            )))
+    }
+
+    fn get_dependents_inner(&self, index: &NodeIndex) -> Result<Vec<NodeIndex>, Error> {
+        Ok(self
+            .graph
+            .neighbors_directed(*index, petgraph::Direction::Outgoing)
+            .collect())
+    }
+}
+
+impl Storage for GraphStorage {
```

Comment:
Do we want tests to ensure this upholds the `Storage` contract with tests? Typically you would, but since we are only using this "real" impl in the `pool` tests (not some fake), maybe we don't need them. Same with the the other trait impls:
```rs
    pub(crate) fn build(
        self,
    ) -> Pool<
        MockDBProvider,
        GraphStorage,
        BasicCollisionManager<NodeIndex>,
        RatioTipGasSelection<NodeIndex>,
    > {
        Pool::new(
            MockDBProvider(self.mock_db),
            GraphStorage::new(GraphConfig {
                max_dependent_txn_count: self.config.max_dependent_txn_count,
            }),
            BasicCollisionManager::new(),
            RatioTipGasSelection::new(),
            self.config,
        )
    }
```

Reply:
Fow now the tests are just high-level on the pool tests should be done on each components. Here is an issue about it : https://github.com/FuelLabs/fuel-core/issues/2196